File: Tests/SentryTests/Protocol/SentrySdkInfo+Equality.m
#import "SentrySdkInfo+Equality.h"
#import <Sentry/Sentry-Swift.h>

@implementation SentrySdkInfo (Equality)

- (BOOL)isEqual:(id _Nullable)object
{
    if (object == self)
        return YES;
    if ([self class] != [object class])
        return NO;

    SentrySdkInfo *otherSdkInfo = (SentrySdkInfo *)object;

    if (![self.name isEqualToString:otherSdkInfo.name]) {
        return NO;
    }

    if (![self.version isEqualToString:otherSdkInfo.version]) {
        return NO;
    }

    if (![[NSSet setWithArray:self.integrations]
            isEqualToSet:[NSSet setWithArray:otherSdkInfo.integrations]]) {
        return NO;
    }

    if (![[NSSet setWithArray:self.features]
            isEqualToSet:[NSSet setWithArray:otherSdkInfo.features]]) {
        return NO;
    }

    if (![[NSSet setWithArray:self.packages]
            isEqualToSet:[NSSet setWithArray:otherSdkInfo.packages]]) {
        return NO;
    }

    if (![self.settings isEqual:otherSdkInfo.settings]) {
        return NO;
    }

    return YES;
}

- (NSUInteger)hash
{
    NSUInteger hash = 17;

    hash = hash * 23 + [self.name hash];
    hash = hash * 23 + [self.version hash];
    hash = hash * 23 + [self.integrations hash];
    hash = hash * 23 + [self.features hash];
    hash = hash * 23 + [self.packages hash];
    hash = hash * 23 + [self.settings hash];

    return hash;
}

@end

File: Tests/SentryTests/Protocol/SentrySDKInfo+Equality.swift
#if SWIFT_PACKAGE
@_spi(Private) @testable import SentrySwift
import _SentryPrivate
import SentryObjCInternal
import SentryTestsObjCHelpers
#else
@testable import Sentry
#endif

extension SentrySdkInfo {
    public static func == (lhs: SentrySdkInfo, rhs: SentrySdkInfo) -> Bool {
        return lhs.name == rhs.name &&
        lhs.version == rhs.version &&
        Set(lhs.integrations) == Set(rhs.integrations) &&
        Set(lhs.features) == Set(rhs.features) &&
        Set(lhs.packages) == Set(rhs.packages) &&
        lhs.settings == rhs.settings
    }
}

#if compiler(>=6.0) && !SWIFT_PACKAGE
extension SentrySdkInfo: @retroactive Equatable { }
#else
extension SentrySdkInfo: Equatable { }
#endif
